src/LinkDotNet.NCronJob/Scheduler/CronScheduler.cs

```diff
@@ -1,47 +1,149 @@
+    private readonly Dictionary<Type, bool> runningJobTypes = [];
+
 
     public CronScheduler(
         JobExecutor jobExecutor,
         CronRegistry registry,
         TimeProvider timeProvider,
-        ILogger<CronScheduler> logger)
+        ConcurrencyConfig concurrencyConfig,
+        ILoggerFactory loggerFactory,
+        IHostApplicationLifetime lifetime)
     {
         this.jobExecutor = jobExecutor;
         this.registry = registry;
         this.timeProvider = timeProvider;
-        this.logger = logger;
+        this.logger = loggerFactory.CreateLogger<CronScheduler>();
+        this.maxDegreeOfParallelism = concurrencyConfig.MaxDegreeOfParallelism;
+        this.semaphore = new SemaphoreSlim(concurrencyConfig.MaxDegreeOfParallelism);
+        this.retryHandler = new RetryHandler(loggerFactory);
+
+        lifetime.ApplicationStopping.Register(() => this.shutdown?.Cancel());
     }
+    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+    {
+        shutdown = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
+        var stopToken = shutdown.Token;
+        stopToken.Register(LogCancellationRequestedInJob);
+        var runningTasks = new List<Task>();
+
+        ScheduleInitialJobs();
+
+        try
+        {
+            while (!stopToken.IsCancellationRequested && jobQueue.Count > 0)
+            {
+                // Remove completed or canceled tasks from the list to avoid memory overflow
+                runningTasks.RemoveAll(t => t.IsCompleted || t.IsFaulted || t.IsCanceled);
+
+                if (jobQueue.TryPeek(out var nextJob, out var nextRunTime) && runningTasks.Count < maxDegreeOfParallelism && CanStartJob(nextJob))
+                {
+                    var delay = nextRunTime - DateTime.UtcNow;
+                    if (delay > TimeSpan.Zero)
+                    {
+                        await Task.Delay(delay, timeProvider, stopToken);
+                    }
+
+                    stopToken.ThrowIfCancellationRequested();
 
-    protected override Task ExecuteAsync(CancellationToken stoppingToken)
+                    // Check if the job supports concurrency
+                    if (!runningJobTypes.TryGetValue(nextJob.Type, out var isRunning) || !isRunning)
+                    {
+                        jobQueue.Dequeue();
+                        runningJobTypes[nextJob.Type] = true;  // Mark as running
+
+                        await semaphore.WaitAsync(stopToken);
+                        var task = Task.Run(async () =>
+                        {
+                            try
+                            {
+                                await ExecuteJob(nextJob, stopToken);
+                            }
+                            finally
+                            {
+                                semaphore.Release();
+                                runningJobTypes[nextJob.Type] = false;  // Mark as not running
```

Comment:
This can lead to race conditions. The dictionary is not thread-safe

Reply:
Noted, I'll make the change.

Reply:
Converted to ConcurrentDictionary